pkcs11/signer.go

```diff
@@ -38,13 +36,33 @@ import (
 	"github.com/theparanoids/crypki/x509cert"
 )
 
-/* Request holds information needed by the collector to fetch the request & process it.
-   It also has a channel on which it waits for the response.
-*/
+// Request holds information needed by the collector to fetch the request & process it.
+// It has multiple channels, one for response & other to notify the worker if the client request times out to stop
+// processing any request from the client.
 type Request struct {
-	pool       sPool                        // pool is a signer pool per identifier from which to fetch the signer
-	identifier string                       // identifier indicates the endpoint for which we are fetching the signer in order to sign it
-	respChan   chan signerWithSignAlgorithm // respChan is the channel where the worker sends the signer once it gets it from the pool
+	pool       sPool          // pool is a signer pool per identifier from which to fetch the signer
+	method     string         // method is type of method the worker needs to call based on the request
+	stop       chan bool      // stop is the channel which is closed when the consumer/client times out or cancels request
+	signerData signerMetadata // signerMetadata is an interface which each request can override & use to sign data
```

Comment:
Functions `getData` and `signData` in `signerMetadata` have the same function arguments and return types, so I'm thinking whether it is possible to combine the `method string` and `signerData signerMetadata` to a unified function var like:
```suggestion

// SignerProcess is a function type to encapsulate the behavior of a request process. 
type SignerProcess (ctx context.Context, signer signerWithSignAlgorithm, pool sPool, data chan []byte, errCh chan error)

type Request struct {
	pool       sPool          // pool is a signer pool per identifier from which to fetch the signer
	stop       chan bool      // stop is the channel which is closed when the consumer/client times out or cancels request
	processData SignerProcess // processData refers to the exact function to request & sign data.
```

In this way, maybe we can get rid of:
```
			case "GetSSHCertSigningKey", "GetX509CACert", "GetBlobSigningPublicKey":
				go w.work.signerData.getData(requestCtx, sResp.signer, w.work.pool, w.work.respChan, w.work.errChan, done)
			case "SignSSHCert", "SignX509Cert", "SignBlob":
				go w.work.signerData.signData(requestCtx, sResp.signer, w.work.pool, w.work.respChan, w.work.errChan, done)
```

Reply:
Yes in a follow up PR, we can only have 1 interface instead of both certsign & signerMetadata as all the signing operation is done by the worker.